### PR TITLE
Property location and jump to location in mouse over popup #370

### PR DESCRIPTION
**Pull request by vrubezhny**
The location link is added to the hover:

Screenshot from 2023-03-02 19-53-46

### CODE REVIEW

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/utils/MarkdownUtils.java [hunk 1]
```diff
@@ -47,12 +52,30 @@ public static String htmlXMLToMarkdown(String description) {
...
+	public static String toLink(String uri, Range range, String message, String title) {
 		StringBuilder link = new StringBuilder();
 		
 		// [Message](http://example.com/ "Title")
 		link.append('[').append(message != null ? message : "This link").append(']');   
 		if (uri != null) {
 			link.append('(').append(uri);
+			if(range != null && (range.getStart() != null || range.getEnd() != null)) {
+				// #L34,1-L35,3
+				StringBuffer selection = new StringBuffer();
```
Comment (angelozerr) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/utils/MarkdownUtils.java:67]: StringBuilder

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/MavenPropertyHoverTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	@AfterEach
+	public void tearDown() throws InterruptedException, ExecutionException {
+		languageService.dispose();
+		languageService = null;
+	}
+	
+	@Test
+	public void testHpverForVariablePropertyDefinedInSameDocument()
+			throws IOException, InterruptedException, ExecutionException, URISyntaxException {
+		DOMDocument document = createDOMDocument("/pom-with-properties-in-parent-for-definition.xml", languageService);
```
Comment (angelozerr) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/MavenPropertyHoverTest.java:48]: I suggest that you use XMLAssert.assertHover from lemminx

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/MavenPropertyHoverTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		Hover hover = languageService.doHover(document, new Position(28, 15), new SharedSettings());
+		assertNotNull(hover);
+		String hoverContents = hover.getContents().getRight().getValue();
+		assertNotNull(hoverContents);
+		assertTrue(hoverContents.contains("anotherProperty"));
+		assertTrue(hoverContents.contains("$"));
+		assertTrue(hoverContents.contains("org.test:child:0.0.1-SNAPSHOT"));
```
Comment (mickaelistria) [inline lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/MavenPropertyHoverTest.java:55]: Can we also check the link is present?

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/utils/MarkdownUtils.java (hunk 1 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/utils/MarkdownUtils.java:67]: Sure. Sorry, content assist inserts StringBuffer as first selection.

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/MavenPropertyHoverTest.java (hunk 2 above)
Comment (angelozerr) [reply lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/MavenPropertyHoverTest.java:48]: See sample at https://github.com/eclipse/lemminx/blob/fe8beb4d9d35339e93befe966efcf5355eee91d7/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/XMLSchemaHoverExtensionsTest.java#L45

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/MavenPropertyHoverTest.java (hunk 3 above)
Comment (vrubezhny) [reply lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/MavenPropertyHoverTest.java]: Yes. Using `XMLAssert.assertHover(...)` solves it.

File: lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/MavenPropertyHoverTest.java (hunk 2 above)
Comment (vrubezhny) [reply lemminx-maven/src/test/java/org/eclipse/lemminx/extensions/maven/participants/hover/MavenPropertyHoverTest.java:50]: OK

**Comment by mickaelistria** [comment]
Thank you!

**Comment by Kiiv** [comment]
So great! Thanks, I miss this so much 🙂

**Comment by angelozerr** [comment]
@vrubezhny I forget to say you that you need to use System.lineSeparator when you write markdown instead of hard coding with\n because If I remember it will not work on Windows os on vscode

**Comment by vrubezhny** [comment]
> @vrubezhny I forget to say you that you need to use System.lineSeparator when you write markdown instead of hard coding with\n because If I remember it will not work on Windows os on vscode

I'll add a PR  to fix this.

@mickaelistria  @angelozerr  What do you think of "double" line break in hover between the rows? (The line spaces seen in the screenshot are made by double newline chars `(\n \n`) - do we really need this extra space in hover?

**Comment by angelozerr** [comment]
> I'll add a PR to fix this.

Thanks, and please use System.lineSeparator in the extected result of test. It will work for you and for CI build if you use \n but not for Windows OS

**Comment by mickaelistria** [comment]
I don't really care at the moment about it. I'd like to see how this looks in actual hover to take a decision here.

**Comment by vrubezhny** [comment]
@mickaelistria Well. It was probably a stupid question... It looks like markdown needs that "double" new line to really add a line break. With having only one new line char the hover looks like the following (everything is placed in one line):
Screenshot from 2023-03-03 14-40-06

**Pull request merged**